### PR TITLE
Axes.inset_axes: enable Axes subclass creation

### LINKED ISSUE #22630: [ENH]: enable passing of projection keyword to Axes.inset_axes
### Problem

Most methods to create an `Axes` instance (e.g. [Figure.add_axes](https://matplotlib.org/stable/api/figure_api.html?highlight=add_axes#matplotlib.figure.Figure.add_axes), [pyplot.subplot](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.subplot.html?highlight=subplot#matplotlib.pyplot.subplot)) allow passing of the `projection` keyword in order to return an instance of a subclass of `Axes`.  It would be useful if [Axes.inset_axes](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.inset_axes.html?highlight=inset_axes#matplotlib.axes.Axes.inset_axes) could do the same.

See also #22608 where there was some agreement that this would be useful, but that was not the right solution.

### Proposed solution

I'm unfamilar with the Matplotlib codebase, but I _think_ the relevant functionality is contained within [Figure._process_projection_requirements](https://github.com/matplotlib/matplotlib/blob/09b17b128440e18cec4a0958dd2dc5b86c9bd146/lib/matplotlib/figure.py#L1524).  So perhaps `inset_axes` could reuse that somehow.

### PR DESCRIPTION
**Pull request by rcomer**
## PR Summary

Closes #22630 

Modify `Axes.inset_axes` so it can accept the `polar`, `projection` and `axes_class` keywords, and therefore return subclasses of `Axes`.  My target use-case is with Cartopy, so we can now adapt [this gallery example](https://scitools.org.uk/cartopy/docs/latest/gallery/scalar_data/waves.html#sphx-glr-gallery-scalar-data-waves-py) to do something like

```python
import cartopy.crs as ccrs
import matplotlib.pyplot as plt
import numpy as np

def sample_data(shape=(73, 145)):
    """Return ``lons``, ``lats`` and ``data`` of some fake data."""
    nlats, nlons = shape
    lats = np.linspace(-np.pi / 2, np.pi / 2, nlats)
    lons = np.linspace(0, 2 * np.pi, nlons)
    lons, lats = np.meshgrid(lons, lats)
    wave = 0.75 * (np.sin(2 * lats) ** 8) * np.cos(4 * lons)
    mean = 0.5 * np.cos(2 * lats) * ((np.sin(2 * lats)) ** 2 + 2)

    lats = np.rad2deg(lats)
    lons = np.rad2deg(lons)
    data = wave + mean

    return lons, lats, data


lons, lats, data = sample_data()

ax = plt.subplot(111, projection=ccrs.PlateCarree())
ax.contourf(lons, lats, data)
ax.coastlines()

axins = ax.inset_axes([0.5, 0.03, 0.47, 0.47], projection=ccrs.PlateCarree())
axins.contourf(lons, lats, data)
axins.set_xlim(-10, 5)
axins.set_ylim(50, 60)
axins.coastlines()

box, lines = ax.indicate_inset_zoom(axins, edgecolor="black")

plt.show()
```

cartopy_inset

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes). _**I'm unsure what extra tests should be added for this.  Advice welcome!**_
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [x] New features are documented, with examples if plot related.
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Comment by rcomer** [comment]
Looks like I have broken [one test](https://github.com/matplotlib/matplotlib/blob/808dbe4a709d286b6dd51c06f111259deaf3f106/lib/matplotlib/tests/test_axes.py#L6691) consistently across platforms.  I will have to come back to this another day.

**Comment by jklymak** [comment]
Inset axes are explicitly children of their parent axes so please don't also add the figure as a parent.  I'm not quite sure what the best way to get a projection into an inset_axes is...

**Comment by greglucas** [comment]
This seems like a useful thing to have available in some form. Might be worth having a look at https://github.com/matplotlib/matplotlib/pull/22060 and whether https://github.com/matplotlib/matplotview would work with GeoAxes too...

Ahh, the axes being a child of the figure or not would actually explain why mpl_toolkit inset_axes warn with `tight_layout()` (because they are added!), but standard inset_axes don't. (xref: #22611)
https://github.com/matplotlib/matplotlib/blob/e75b5291108bf0f4d15785612c7a7899d07ea9a2/lib/mpl_toolkits/axes_grid1/inset_locator.py#L294

The easiest way might be to allow `axes_class` to be passed in to the `inset_axes()` method, similar to mpl_toolkits and figure's add_axes methods.
https://github.com/matplotlib/matplotlib/blob/e75b5291108bf0f4d15785612c7a7899d07ea9a2/lib/mpl_toolkits/axes_grid1/inset_locator.py#L380-L385

However, there might be some more surgery needed outside of AxesGrid to support the projection or other keyword args.
https://github.com/matplotlib/matplotlib/blob/09b17b128440e18cec4a0958dd2dc5b86c9bd146/lib/matplotlib/figure.py#L535-L538

**Comment by rcomer** [comment]
Thanks for the feedback.  I did have a hunch that it couldn't really be this simple!

> Inset axes are explicitly children of their parent axes so please don't also add the figure as a parent.

I'm afraid I don't have the background knowledge to understand this.  Is there somewhere I can read up on the matplotlib parent/child relationships and why they matter?  I guess the comment about `tight_layout` gives a clue.

> matplotlibview

Looks useful - thanks!  I will have a play with that at some point.

> The easiest way might be to allow axes_class to be passed in to the inset_axes() method, similar to mpl_toolkits and figure's add_axes methods.

I'm not so keen on that API as it seems inconsistent with how the user is usually encouraged to create their Axes instances.  But I guess it was done in those other functions for a reason.

Anyway it seems that whatever the answer is, it's not this.  So for now I will just close this.

**Comment by jklymak** [comment]
@rcomer, maybe convert this to an issue?   I think it is reasonable to ask that inset_axes accepts projections, I'm just not sure off the top of my head how to do it (I have limited Matplotlib (and internet) bandwidth at the moment.

**Comment by rcomer** [comment]
> maybe convert this to an issue?

Issue now at #22630.  Thanks!

**Comment by jklymak** [comment]
@rcomer, did this not work?   It looks like something similar _should_ work, and it is certainly reasonable functionality...

**Comment by rcomer** [comment]
@jklymak yes the CI all seemed fine and it works for my Cartopy example.  However at #22630 @mylasem said they were working on it, so I thought I would leave it to them for now.  I'd be happy to pick this up again at some point if @mylasem's version doesn't land for whatever reason.

**Comment by jklymak** [comment]
In general, Matplotlib does not have a "reserve an issue" policy, and this PR predates @mylasem 's comment.  I think they should work within the context of this PR if at all possible.

**Comment by rcomer** [comment]
OK, since @mylasem has not commented further and I had some time, I have:

* Added some basic tests to verify that the expected type is returned for our three methods of specifying it.
* Updated the `Axes.inset_axes` docstring with the keyword descriptions - this was a straight copy/paste from the `Figure.add_axes` docstring, and then added the word "inset".
* Updated the what's new entry (and the PR title and OP) to reflect that we're not going via `Figure.add_axes`.

File: lib/matplotlib/tests/test_axes.py
Comment (rcomer) [inline lib/matplotlib/tests/test_axes.py:6618]: I imported the class within the test as I noticed another test in this module does that.  If it's not necessary, then I think we could parametrize these three tests into one.

File: lib/matplotlib/tests/test_axes.py
Comment (jklymak) [inline lib/matplotlib/tests/test_axes.py:6622]: This test seems good, but I do wonder if we actually need an image test here to make sure all the drawing gets through properly?  Otherwise this looks great.

Comment (rcomer) [reply lib/matplotlib/tests/test_axes.py:6622]: @jklymak thanks.  I have to admit I've never used matplotlib's own projections before, so will have to do some reading and have a think.  If you happen to know of any specific examples I could adapt, that would be great.

Comment (rcomer) [reply lib/matplotlib/tests/test_axes.py:6623]: @jklymak is something like [this](https://github.com/matplotlib/matplotlib/pull/22608/commits/452622505dd72a5ba6c6745e33772bcf63fc5ea5) sufficient?  From what I can tell, the zoom type plot won't work with polar, because the xlim/ylim don't correspond to a box.  But a user may have other reasons to want an inset plot.

File: lib/matplotlib/tests/test_axes.py
Comment (tacaswell) [inline lib/matplotlib/tests/test_axes.py:6617]:
```suggestion
@image_comparison(['inset_polar.png'], remove_text=True, style='mpl20')
```

But this will also require regenerating the test image.

File: lib/matplotlib/tests/test_axes.py
Comment (tacaswell) [reply lib/matplotlib/tests/test_axes.py:6619]: I do not see any reason to do this, I think pulling all of the imports to the top is OK.

**Review by tacaswell — APPROVED** [review]
I am :+1: with this going in as-is. 


Moving to the mpl20 style would be good as would refactoring the test, but neither is worth holding up the PR over.

**Comment by rcomer** [comment]
Thanks @tacaswell, I am happy to follow those up tomorrow when I am back at my keyboard.  Should I also squash the commits down to one?

**Comment by tacaswell** [comment]
> Should I also squash the commits down to one?

Yes please (and if you replace the image CI without squashing CI will fail).

File: doc/users/next_whats_new/inset_axes_improvements.rst
Comment (QuLogic) [inline doc/users/next_whats_new/inset_axes_improvements.rst:5]: Argument names should use italics.
```suggestion
`matplotlib.axes.Axes.inset_axes` now accepts the *projection*, *polar* and
*axes_class* keyword arguments, so that subclasses of `matplotlib.axes.Axes` may
```

File: lib/matplotlib/axes/_axes.py
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py:336]: We can't remove `polar` because it comes through the internal meothd, but I wonder if we want to not document it, since it duplicates `projection`? @timhoffm

I recall discussing it, and it couldn't really be dropped from normal since it was so widely used, but since this is new, we don't have to worry about existing users.

File: lib/matplotlib/tests/test_axes.py
Comment (rcomer) [reply lib/matplotlib/tests/test_axes.py]: done

File: lib/matplotlib/tests/test_axes.py
Comment (rcomer) [reply lib/matplotlib/tests/test_axes.py]: I moved the imports to the top, but in the end couldn't convince myself it was worth parametrizing the two small tests that are still the same: since they're so small there isn't _much_ repetition going on and I suspect readability is better as they are than with a parametrized version.

File: lib/matplotlib/tests/test_axes.py
Comment (tacaswell) [inline lib/matplotlib/tests/test_axes.py:6621]:
```suggestion
    axins = ax.inset_axes([0.5, 0.1, 0.45, 0.45], projection='polar')
```

given the above discussion?

Comment (rcomer) [reply lib/matplotlib/tests/test_axes.py:6621]: Would that also mean that [this one](https://github.com/matplotlib/matplotlib/blob/80e672e0700fa8a268aed1bdefedbd0e493e91a7/lib/matplotlib/tests/test_axes.py#L6631-L6635) is redundant and should be removed?

File: lib/matplotlib/axes/_axes.py
Comment (timhoffm) [reply lib/matplotlib/axes/_axes.py:336]: Unfortunately, `polar` is a redundant API. OTOH `polar` is widely used and we have not even discouraged its use, I consider it proper first-class API. One can regard it as a shortcut.
Overall, it's there and it's valid to be used. Therefore I recommend to also document it.

Comment (rcomer) [reply lib/matplotlib/axes/_axes.py:336]: I’m not sure where to go from here as there are differing opinions.  FWIW I think it’s intuitive that the various axes creation functions should take the same keywords.  So if I were used to using `polar` with the other functions I would expect to use it with this one too.  I’m not even sure whether I would check the docstring before trying it!

Comment (timhoffm) [reply lib/matplotlib/axes/_axes.py:336]: Picking up my API authority hat, I decide that the present consistent way is the desired one. If we want to do something about `polar`, that'd need to be a separate discussion and a global change (but I advise against that).

**Review by timhoffm — APPROVED** [review]

**Pull request merged**
